compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -427,6 +427,8 @@ pub(super) fn explicit_predicates_of<'tcx>(
             //   supertrait).
             if let ty::Alias(ty::Projection, projection) = ty.kind() {
                 projection.substs == trait_identity_substs
+                    // FIXME(return_type_notation): This check should be more robust
+                    && !tcx.is_impl_trait_in_trait(projection.def_id)
```

Comment:
We can encounter precisely two types of items for the `projection.def_id` here:

1. Associated types/consts
2. return-position impl trait in trait, when using RTN

If the return type is not an RPITIT, then it's always going to be an associated type or const, and the next line will never ICE. If the return type *is* an RPITIT, we shouldn't be calling `associated_item` on it, since that'll ICE.

When I mentioned that it should be more robust, I wasn't saying it was incomplete, I just don't really like that we have to treat associated items and RPITITs differently here (but this is a consequence of how they're currently implemented, and most code doesn't care about this distinction anyways).